app/src/main/java/sdp/moneyrun/Game.java

```diff
@@ -68,37 +109,164 @@ public Game(String gameId,
-        Intent endGameIntent = new Intent(currentActivity, EndGameActivity.class);
-        ArrayList<Integer> collectedCoinsValues = new ArrayList<>();
-        for (int i = 0; i < collectedCoins.size(); ++i) {
-            collectedCoinsValues.add(collectedCoins.get(i).getValue());
+
+    /**
+     * Adds the GameData to the database if not already present
+     * @return the Id of this game in the DB
+     */
+    public String addToDB(){
+        if(!hasBeenAdded) {
+            DatabaseReference openGames = rootReference.child("open_games");
+            id = openGames.push().getKey();
+            openGames.child(id).setValue(gameDbData);
+            linkAttributesToDB();
+            hasBeenAdded = true;
+            return id;
         }
-        endGameIntent.putExtra("collectedCoins", collectedCoinsValues);
-        endGameIntent.putExtra("playerId", playerId);
-        currentActivity.startActivity(endGameIntent);
-        currentActivity.finish();
+        return getGameId();
     }
 
-    public String getGameId() {
-        return gameId;
+
+
+    
+    /**
+     * Links pertinent attributes to the DB instance corresponding to its ID.
+     * For now the only pertitent attribute is the player List
+     */
+    private void linkAttributesToDB(){
+        DatabaseReference gameRef = rootReference.child("open_games").child(id);
+        gameRef.child("players").addValueEventListener(new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot snapshot) {
+                GenericTypeIndicator<List<Player>> t = new GenericTypeIndicator<List<Player>>() {
+                };
+                List<Player> newData = snapshot.getValue(t);
+                gameDbData.setPlayers(newData);
+            }
+
+            @Override
+            public void onCancelled(@NonNull DatabaseError error) {
+                Log.e(TAG, "failed " + error.getMessage());
+            }
+        });
     }
 
-    public String getName() {
-        return name;
+    /**
+     * Gets a Serialized Game from the DB in an asynchronous manner
+     * @param id ID of the Game to retrieve
+     * @return a Task containing the serialized Game
+     * @throws NoSuchElementException if there is no Game with this ID in the DB
+     */
+    public static Task<DataSnapshot> getGameDataSnapshot(String id) throws NoSuchElementException{
+        if(id == null){throw new IllegalArgumentException();}
+        DatabaseReference gamesRef = FirebaseDatabase.getInstance()
+                .getReference().child("open_games");
+
+        Task<DataSnapshot> returnSnap = gamesRef.child(id).get();
+
+        returnSnap.addOnCompleteListener(task -> {
+            if (task.isSuccessful()) {
+                if(task.getResult().child("name").getValue(String.class) == null){
+                    throw new NoSuchElementException("There is no game with ID : " + id + " in the DB");
```

Comment:
I think it's dangerous to have such a mechanism of an asynchronous exception throw. It would be better IMO if the caller deal with the queried object being non existent

Reply:
good point, should I just add in the javadoc that : "@returns a task containing a Null value if the game is not present in the DB" and eliminate this throws? Because i dont see how to deal with it otherwose since we wont know if its null until it is complete

Reply:
yea, nevermind I just removed the whole block and added a warning in @returns that said it could be null, i dont know how else to warn that it might become a nullpointer somewhere :/